src/python/WMCore/WorkQueue/WorkQueueBackend.py

```diff
@@ -458,87 +464,107 @@ def availableWork(self, thresholds, siteJobCounts, team=None,
         # Fetch workqueue elements in slices, using the CouchDB "limit" and "skip"
         # options for couch views. Conditions to stop this loop are:
         #  a) have a hard stop at 50k+1 (we might have to make this configurable)
-        #  b) stop as soon as an empty slice is returned by Couch (thus all docs have
-        #     already been retrieve)
+        #  b) stop once total_rows is reached
         #  c) or, once "numElems" elements have been accepted
-        numSkip = 0
-        breakOut = False
-        while True:
-            if breakOut:
-                # then we have reached the maximum number of elements to be accepted
-                break
-            self.logger.info("  with limit docs: %s, and skip first %s docs", docsSliceSize, numSkip)
-            options['skip'] = numSkip
+        numSlices = ceil(numAvail / docsSliceSize)
+        # limit it to the first 50k, in case there are tons of work in the system
+        numSlices = min(numSlices, int(50001 / docsSliceSize))
```

Comment:
Now Unified is controlling the amount of work assigned, so it might be that we will never go beyond 100k available elements. From the CouchDB experts, skipping large amount of rows (whatever is the definition or large here) is discouraged, so yet another reason to having a max limit here. This is also a safe mechanism to ensure that the system remains responsive to whatever global workqueue scenario we get exposed to.

Let me check if I can easily make this one configurable as well.